include/onnxruntime/core/session/onnxruntime_cxx_inline.h

```diff
@@ -915,6 +915,13 @@ inline std::vector<Value> SessionImpl<T>::Run(const RunOptions& run_options, con
   return output_values;
 }
 
+template <typename T>
+inline void SessionImpl<T>::RunAsync(const RunOptions& run_options, const char* const* input_names, const Value* input_values, size_t input_count,
+                                     const char* const* output_names, size_t output_count, RunAsyncCallbackFn callback) {
+  auto ort_input_values = reinterpret_cast<const OrtValue* const*>(input_values);
+  ThrowOnError(GetApi().RunAsync(this->p_, run_options, input_names, ort_input_values, input_count, output_names, output_count, callback));
```

Comment:
May need design decision of how to deal with exceptions within async context. could be one of the following:
1. RunAsync() won't throw exceptions. Callback is guaranteed to be called and any error will be included in the callback's OrtStatus parameter
2. RunAsync() may throw exceptions.  Callback is still guaranteed to be called and any error will be included in the callback's OrtStatus parameter.
3. RunAsync() may throw exceptions. Callback will not be called if RunAsync() throws; otherwise it is guaranteed to be called and if any error occurred after RunAsync() returns, it will be included in the callback's OrtStatus parameter.

I strongly recommend not to throw exceptions in async context because today's C++ language feature does not support a unified exception handling mechanism within both sync and async context. ( C#, JavaScript and some other languages support this )

Reply:
By the way, in C API RunAsync() returns void, and we don't expect any exceptions thrown in C API. so this function should simply never throw.

Reply:
RunAsync might throw, and when that happens, the callback will not be called.

Reply:
No exceptions should be thrown at the C API layer. This is the reason we've OrtStatus everywhere. RunAsync C API should follow the same pattern. There's no way for the caller to know if the async call failed due to exception or something else; hence the callback should be called in all scenarios to communicate this information.